### PR TITLE
More efficient handling of transparent states

### PR DESCRIPTION
**Pull request by SkySkimmer**
Hopefully speeds up the cases in https://github.com/coq/coq/issues/18168 / removes the slowdown from https://github.com/coq/coq/pull/13448

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
(bench queued)

File: clib/hMap.ml [hunk 1]
```diff
@@ -38,6 +38,8 @@ struct
   let is_empty = Int.Map.is_empty
 
   let mem x s =
+    if Int.Map.is_empty s then false
```
Comment (SkySkimmer) [inline clib/hMap.ml:41]: So this tiny change is enough to remove the slowdown in hott/category theory
but presumably if they add anything to the transparent state that won't work anymore

**Comment by RalfJung** [comment]
I've run our benchmark to compare coq.dev and this branch, [here](https://coq-speed.mpi-sws.org/d/1QE_dqjiz/coq-compare?orgId=1&var-project=iris&var-branch1=ci%2Ftiming&var-commit1=2a82e4d93f55b68077ede24821e164a6e80dc022&var-config1=build-coq.dev&var-branch2=ci%2Ftiming&var-commit2=7429a2500c8edb75c291414b7e5409df23a102ec&var-config2=build-coq.dev&var-metric=instructions&var-group=%28.%2A%29) are the results. It's definitely a big improvement! It doesn't completely un-do the regression, but reduces it by a lot. Looking at the two files I mentioned in the issue:
- proph_erasure.v: Coq 8.17: 105 Billion (instructions), with this PR: 112 Billion; that's a 7% slowdown
- lang.v: Coq 8.17: 72 Billion, with this PR: 77 Billion; that's a 7% slowdown

For Iris as a whole, with this PR Coq is around 2.3% slower than Coq 8.17 was.

**Pull request closed without merging**